[VPC] Add authentication validation for events

nasark opened the pull request:
- added auth check in event catcher and validation of events API service key credentials


Auth validation:
```
[----] I, [2022-03-23T15:00:14.548900 #42007:52f8]  INFO -- evm: MIQ(MiqAeEngine.deliver) Delivering {:event_type=>"ems_auth_valid", "MiqEvent::miq_event"=>937, :miq_event_id=>937, "EventStream::event_stream"=>937, :event_stream_id=>937} for object [ManageIQ::Providers::IbmCloud::VPC::CloudManager.154] with state [] to Automate
[----] I, [2022-03-23T15:00:14.643737 #42007:52f8]  INFO -- evm: MIQ(MiqEvent#process_evm_event) Event Raised [ems_auth_valid]
[----] I, [2022-03-23T15:00:14.650975 #42007:52f8]  INFO -- evm: MIQ(MiqEvent#process_evm_event) Alert for Event [ems_auth_valid]
[----] I, [2022-03-23T15:00:14.651039 #42007:52f8]  INFO -- evm: MIQ(MiqAlert.evaluate_alerts) [ems_auth_valid] Target: ManageIQ::Providers::IbmCloud::VPC::CloudManager Name: [na], Id: [154]
[----] I, [2022-03-23T15:00:14.653466 #42007:52f8]  INFO -- evm: MIQ(MiqQueue#delivered) Message id: [44521], State: [ok], Delivered in [0.107928] seconds
```

Targeted refresh triggered via events afterwards:
```
[----] I, [2022-03-23T15:04:21.705848 #42007:52f8]  INFO -- evm: MIQ(ManageIQ::Providers::IbmCloud::VPC::CloudManager::Refresher#refresh) Refreshing all targets...
[----] I, [2022-03-23T15:04:21.705899 #42007:52f8]  INFO -- evm: MIQ(ManageIQ::Providers::IbmCloud::VPC::CloudManager::Refresher#refresh) EMS: [na], id: [154] Refreshing targets for EMS...
[----] I, [2022-03-23T15:04:21.705945 #42007:52f8]  INFO -- evm: MIQ(ManageIQ::Providers::IbmCloud::VPC::CloudManager::Refresher#refresh) EMS: [na], id: [154]   InventoryRefresh::TargetCollection [Collection of 1 targets] id [Collection of targets with id: [{:ems_ref=>nil}]]
[----] I, [2022-03-23T15:04:21.783921 #42007:52f8]  INFO -- evm: MIQ(ManageIQ::Providers::IbmCloud::VPC::CloudManager::Refresher#refresh_targets_for_ems) EMS: [na], id: [154] Refreshing target InventoryRefresh::TargetCollection [Collection of 1 targets] id [Collection of targets with id: [{:ems_ref=>nil}]]...
[----] I, [2022-03-23T15:04:21.793887 #42007:52f8]  INFO -- evm: EMS: [na], id: [154] Saving EMS Inventory...
[----] I, [2022-03-23T15:04:21.858086 #42007:52f8]  INFO -- evm: EMS: [na], id: [154] Saving EMS Inventory...Complete
[----] I, [2022-03-23T15:04:21.858210 #42007:52f8]  INFO -- evm: MIQ(ManageIQ::Providers::IbmCloud::VPC::CloudManager::Refresher#refresh_targets_for_ems) EMS: [na], id: [154] Refreshing target InventoryRefresh::TargetCollection [Collection of 1 targets] id [Collection of targets with id: [{:ems_ref=>nil}]]...Complete
[----] I, [2022-03-23T15:04:21.858269 #42007:52f8]  INFO -- evm: MIQ(ManageIQ::Providers::IbmCloud::VPC::CloudManager::Refresher#refresh) EMS: [na], id: [154] Refreshing targets for EMS...Complete - Timings {:collect_inventory_for_targets=>0.07789492607116699, :parse_targeted_inventory=>0.0003268718719482422, :save_inventory=>0.07385396957397461, :publish_inventory=>1.6927719116210938e-05, :ems_refresh=>0.15224814414978027}
[----] I, [2022-03-23T15:04:21.883544 #42007:52f8]  INFO -- evm: MIQ(ManageIQ::Providers::IbmCloud::VPC::CloudManager::Refresher#refresh) Refreshing all targets...Complete
[----] I, [2022-03-23T15:04:21.883721 #42007:52f8]  INFO -- evm: MIQ(MiqQueue#delivered) Message id: [44556], State: [ok], Delivered in [0.182403] seconds
```

@miq-bot assign @agrare 
@miq-bot add_label enhancement

agrare commented on app/models/manageiq/providers/ibm_cloud/vpc/manager_mixin.rb:
```diff
@@ -19,11 +19,21 @@ def connect(options = {})
   # @param options [Hash] Connection options.
-  # @return [ManageIQ::Providers::IbmCloud::CloudTools::Vpc]
+  # @return [ManageIQ::Providers::IbmCloud::CloudTools::Vpc] or [Boolean]
   def verify_credentials(_auth_type = nil, options = {})
```
If you're going to use the auth_type passed in we should drop the leading `_` as that is an indication to linters (rubocop) that it is okay that the argument is unused.

agrare commented on app/models/manageiq/providers/ibm_cloud/vpc/manager_mixin.rb:
```diff
@@ -19,11 +19,21 @@ def connect(options = {})
+    when :events
+      verify_events_credentials
+    else
+      connect(options).authenticator
```
For symmetry maybe we can pull this out into a `verify_default_credentials` method

agrare commented on app/models/manageiq/providers/ibm_cloud/vpc/manager_mixin.rb:
```diff
@@ -19,11 +19,21 @@ def connect(options = {})
+
+  def verify_events_credentials
+    service_key = authentication_key("events")
+    connect.resource.controller.collection(:list_resource_keys).find {|key| key.dig(:credentials, :service_key) == service_key}.present?
```
You should be able to use `.any? {}` instead of `.find {}.present?`

agrare commented on app/models/manageiq/providers/ibm_cloud/vpc/manager_mixin.rb:
```diff
@@ -19,11 +19,21 @@ def connect(options = {})
+
+  def verify_events_credentials
+    service_key = authentication_key("events")
+    connect.resource.controller.collection(:list_resource_keys).find {|key| key.dig(:credentials, :service_key) == service_key}.present?
```
Also minor style issue
```suggestion
    connect.resource.controller.collection(:list_resource_keys).find { |key| key.dig(:credentials, :service_key) == service_key }.present?
```

agrare replied:
Do we not need to pass `options` into connect here like we do up [[here]](https://github.com/ManageIQ/manageiq-providers-ibm_cloud/pull/359/files#diff-9c1c1fe8a8cf1522f3f23dad6743d18f6c423658360cd877f66c9cb7ac3a1753R30)?  Granted I don't know what is typically in options at this point

nasark replied:
@agrare `options` from what I can tell is an empty hash, should we remove it?

agrare replied:
I think lets pass it in to be consistent with the existing `connect()`

nasark commented:
@agrare Thanks, addressed the feedback. Just need a verdict on the `options` (empty hash) being passed into `connect`
https://github.com/ManageIQ/manageiq-providers-ibm_cloud/pull/359/files#diff-9c1c1fe8a8cf1522f3f23dad6743d18f6c423658360cd877f66c9cb7ac3a1753R35

passed to:
https://github.com/ManageIQ/manageiq-providers-ibm_cloud/blob/371a09b1c442caf5b24ad20ef65859aa0f314147/app/models/manageiq/providers/ibm_cloud/vpc/manager_mixin.rb#L16

agrare commented on app/models/manageiq/providers/ibm_cloud/vpc/manager_mixin.rb:
```diff
@@ -19,13 +19,27 @@ def connect(options = {})
 
   # Same as calling connect.
-  # @param _auth_type [nil] Not used
+  # @param _auth_type [nil]
```
We should update the name of the param in this comment too

agrare commented on app/models/manageiq/providers/ibm_cloud/vpc/cloud_manager/event_catcher.rb:
```diff
@@ -1,3 +1,9 @@
+
+  def self.all_valid_ems_in_zone
+    super.select do |ems|
+      ems.authentication_key("events").present?
```
In a follow-up I think we can also check `ems.authentication_status_ok?(:events)` which would cover the case where someone added an api service key but it isn't valid

nasark replied:
Sure, will do

Fryguy commented:
Backported to `najdorf` in commit 0a5e1e6a8b38b7e340dc587cc97f7e3cb2080dfe.

```text
commit 0a5e1e6a8b38b7e340dc587cc97f7e3cb2080dfe
Author: Adam Grare <adam@grare.com>
Date:   Thu Mar 24 09:37:09 2022 -0400

    Merge pull request #359 from nasark/events_worker_auth_validation
    
    [VPC] Add authentication validation for events
    
    (cherry picked from commit f60fe6e44e526f94ad77127f7043a1eab4fb8136)
```
